File: internal/ui/spinner.go
package ui

import (
	"fmt"
	"os"
	"sync"
	"time"
)

const frameEvery = 80 * time.Millisecond

type Spinner struct {
	theme *Theme
	live  bool
	start time.Time

	mu    sync.Mutex
	label string

	done chan struct{}
	wg   sync.WaitGroup
}

func (t *Theme) NewSpinner() *Spinner {
	return &Spinner{
		theme: t,
		live:  IsTerminal(os.Stderr),
		label: "starting",
		done:  make(chan struct{}),
	}
}

func (s *Spinner) Start() {
	s.start = time.Now()
	if !s.live {
		return
	}
	fmt.Fprint(os.Stderr, cursorHide)
	s.wg.Add(1)
	go func() {
		defer s.wg.Done()
		ticker := time.NewTicker(frameEvery)
		defer ticker.Stop()
		for {
			select {
			case <-s.done:
				return
			case <-ticker.C:
				s.draw()
			}
		}
	}()
	s.draw()
}

func (s *Spinner) Status(text string) {
	s.mu.Lock()
	s.label = text
	s.mu.Unlock()
	if s.live {
		s.draw()
		return
	}
	fmt.Fprintf(os.Stderr, "%s %s\n", s.theme.Glyph.Dot, text)
}

func (s *Spinner) draw() {
	elapsed := time.Since(s.start)
	frames := s.theme.spinner
	frame := frames[int(elapsed/frameEvery)%len(frames)]

	s.mu.Lock()
	label := s.label
	s.mu.Unlock()

	fmt.Fprintf(os.Stderr, "%s%s %s %s",
		clearLine,
		s.theme.Clay(frame),
		s.theme.Dim(label),
		s.theme.Dim(fmt.Sprintf("%.1fs", elapsed.Seconds())),
	)
}

func (s *Spinner) Stop() {
	if !s.live {
		return
	}
	select {
	case <-s.done:
		return
	default:
		close(s.done)
	}
	s.wg.Wait()
	fmt.Fprint(os.Stderr, clearLine+cursorShow)
}

File: src/ui/spinner.ts
import { clearLine, cursorHide, cursorShow, isTerminal, type Theme } from "./theme";

const frameEvery = 80;

export class Spinner {
  private readonly live: boolean;
  private label = "starting";
  private started = Date.now();
  private timer: NodeJS.Timeout | undefined;

  constructor(private readonly theme: Theme) {
    this.live = isTerminal(process.stderr);
  }

  start(): void {
    this.started = Date.now();
    if (!this.live) {
      return;
    }
    process.stderr.write(cursorHide);
    this.timer = setInterval(() => this.draw(), frameEvery);
    this.timer.unref();
    this.draw();
  }

  status(text: string): void {
    this.label = text;
    if (this.live) {
      this.draw();
      return;
    }
    process.stderr.write(`${this.theme.glyph.dot} ${text}\n`);
  }

  private draw(): void {
    const elapsed = Date.now() - this.started;
    const frames = this.theme.frames;
    const frame = frames[Math.floor(elapsed / frameEvery) % frames.length]!;
    process.stderr.write(
      `${clearLine}${this.theme.clay(frame)} ${this.theme.dim(this.label)} ${this.theme.dim(
        `${(elapsed / 1000).toFixed(1)}s`,
      )}`,
    );
  }

  stop(): void {
    if (!this.live || this.timer === undefined) {
      return;
    }
    clearInterval(this.timer);
    this.timer = undefined;
    process.stderr.write(clearLine + cursorShow);
  }
}
